src/uct/ib/rc/accel/rc_mlx5_common.h

```diff
@@ -219,6 +180,84 @@ uct_rc_mlx5_iface_common_poll_rx(uct_rc_mlx5_iface_common_t *mlx5_common_iface,
+static UCS_F_ALWAYS_INLINE void
+uct_rc_mlx5_iface_common_am_handler(uct_rc_mlx5_iface_common_t *mlx5_iface,
+                                    uct_rc_iface_t *rc_iface,
+                                    struct mlx5_cqe64 *cqe,
+                                    uct_ib_mlx5_srq_seg_t *seg,
+                                    unsigned byte_len)
+{
+    uct_rc_hdr_t *hdr;
+    uint16_t wqe_ctr;
+    uct_ib_iface_recv_desc_t *desc;
+    uct_rc_iface_ops_t *rc_ops;
+    unsigned flags;
+    uint32_t qp_num;
+    ucs_status_t status;
+
+    wqe_ctr = ntohs(cqe->wqe_counter);
+    desc    = seg->srq.desc;
+
+    /* Get a pointer to AM header (after which comes the payload)
+     * Support cases of inline scatter by pointing directly to CQE.
+     */
+    hdr = uct_rc_mlx5_iface_common_data(mlx5_iface, rc_iface, cqe, desc,
+                                        byte_len, &flags);
+
+    uct_ib_mlx5_log_rx(&rc_iface->super, IBV_QPT_RC, cqe, hdr,
+                       uct_rc_ep_am_packet_dump);
+
+    if (ucs_unlikely(hdr->am_id & UCT_RC_EP_FC_MASK)) {
+        qp_num = ntohl(cqe->sop_drop_qpn) & UCS_MASK(UCT_IB_QPN_ORDER);
+        rc_ops = ucs_derived_of(rc_iface->super.ops, uct_rc_iface_ops_t);
+
+        status = rc_ops->fc_handler(rc_iface, qp_num, hdr, byte_len - sizeof(*hdr),
+                                    cqe->imm_inval_pkey, cqe->slid, flags);
+    } else {
+        status = uct_iface_invoke_am(&rc_iface->super.super, hdr->am_id,
+                                     hdr + 1, byte_len - sizeof(*hdr),
+                                     flags);
+    }
+
+    uct_rc_mlx5_iface_common_desc_free(mlx5_iface, rc_iface, seg, wqe_ctr, status,
+                                       rc_iface->super.config.rx_headroom_offset,
+                                       &rc_iface->super.release_desc);
+}
+
+static UCS_F_ALWAYS_INLINE unsigned
+uct_rc_mlx5_iface_common_poll_rx(uct_rc_mlx5_iface_common_t *mlx5_common_iface,
+                                 uct_rc_iface_t *rc_iface)
+{
+    uct_ib_mlx5_srq_seg_t *seg;
+    struct mlx5_cqe64 *cqe;
+    unsigned byte_len;
+    uint16_t max_batch;
+    unsigned count;
+
+    ucs_assert(uct_ib_mlx5_srq_get_wqe(&mlx5_common_iface->rx.srq,
+                                       mlx5_common_iface->rx.srq.mask)->srq.next_wqe_index == 0);
+
+    /* Prefetch the descriptor if it was scheduled */
+    uct_rc_mlx5_srq_prefetch_first(mlx5_common_iface);
+
+    cqe = uct_ib_mlx5_poll_cq(&rc_iface->super, &mlx5_common_iface->rx.cq);
+    if (cqe == NULL) {
+        /* If no CQE - post receives */
+        count = 0;
+        goto done;
+    }
+
+    ucs_memory_cpu_load_fence();
+    UCS_STATS_UPDATE_COUNTER(rc_iface->stats, UCT_RC_IFACE_STAT_RX_COMPLETION, 1);
+
+    byte_len = ntohl(cqe->byte_cnt);
+    seg      = uct_ib_mlx5_srq_get_wqe(&mlx5_common_iface->rx.srq,
+                                       ntohs(cqe->wqe_counter));
+
+    uct_rc_mlx5_iface_common_am_handler(mlx5_common_iface, rc_iface, cqe,
```

Comment:
what is the reason for moving this to separate function?

Reply:
Tag RX poll function will invoke it as well when NO_TAG message arrives

Reply:
will there be separate CQ for tags?

Reply:
CQ will be the same, but the progress function will be different, like with verbs

Reply:
seems like code duplication.. maybe we can have on progress function which have a "boolean" parameter, resolved in compile time, which tells if TM CQE types are to be checked?